### ISSUE TITLE
Support rebuild with tablets

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#17575]
We need to support `nodetool rebuild` with tablets.

Currently, we only support the legacy repair api for tablets, but not the rebuild api
so scylla-nodetool may instead use the repair api to implement rebuild.

Refs https://github.com/scylladb/scylladb/issues/16128

**Comment by mykaul** [comment]
@bhalevy - 
I assume - but please confirm - that this is a release blocker for 6.0?

**Comment by bhalevy** [comment]
> @bhalevy - I assume - but please confirm - that this is a release blocker for 6.0?

Yes.  Per our decision to support adding a new DC with tablets, this became a release blocker.

**Comment by xemul** [comment]
@bhalevy , isn't f3bd7e9b8c610851dad07838c7bc43e99e1085c7 doing what we need?

`api/storage_service.cc`
```c++
    ss::rebuild.set(r, [&ss](std::unique_ptr<http::request> req) {
        auto source_dc = req->get_query_param("source_dc");
        apilog.info("rebuild: source_dc={}", source_dc);
        return ss.local().rebuild(std::move(source_dc)).then([] {
            return make_ready_future<json::json_return_type>(json_void());
        });
    });
```
then
`service/storage_service.cc`
```c++
future<> storage_service::rebuild(sstring source_dc) {
    return run_with_api_lock(sstring("rebuild"), [source_dc] (storage_service& ss) -> future<> {
        ss.check_ability_to_perform_topology_operation("rebuild");
        if (ss.raft_topology_change_enabled()) {
            co_await ss.raft_rebuild(source_dc);
```
and so on, so what `nodetool rebuild` support is missing?

**Comment by xemul** [comment]
There's even a test for that :) 517f6bfa8a9ffc1a0a8afb55b7236507acbc0657

**Comment by tchaikov** [comment]
@xemul could you check if the test failure https://jenkins.scylladb.com/job/scylla-master/job/byo/job/dtest-byo/186/testReport/rebuild_test/TestRebuild/FullDtest___full_split003___test_simple_rebuild/ is expected? the test tried to rebuild a second node from an existing one.

**Comment by xemul** [comment]
@tchaikov , the test does this:

> session.execute("ALTER KEYSPACE ks WITH REPLICATION = {'class':'NetworkTopologyStrategy', 'dc1':1, 'dc2':1};")

trying to make data be replicated to another dc as well. It's not supposed to work until #17116 , so the test worth marking with this issue or !tablets until that

**Comment by tchaikov** [comment]
thank you, @xemul . i am updating the spreadsheet tracking the test failures accordingly.

**Comment by xemul** [comment]
> There's even a test for that :) [517f6bf](https://github.com/scylladb/scylladb/commit/517f6bfa8a9ffc1a0a8afb55b7236507acbc0657)

OK, so

1. test doesn't create user tables
2. test doesn't enable tablets
3. even if creating a table with tablets rebuild succeeds, but no traces of rebuilding the test table are found in logs :(

**Comment by bhalevy** [comment]
@xemul what's the status of this issue?
I had a vague memory that there's an open PR to fix this issue, but I probably confused it with some other PR, since I can't find any.
Is there one?

**Comment by xemul** [comment]
> @xemul what's the status of this issue?

I believe that the current status is: we don't know if we need it at all, because rebuild is needed for "adding new DC" which is tracked by #17572

**Comment by bhalevy** [comment]
> > [@xemul](https://github.com/xemul) what's the status of this issue?
> 
> I believe that the current status is: we don't know if we need it at all, because rebuild is needed for "adding new DC" which is tracked by [#17572](https://github.com/scylladb/scylladb/issues/17572)

removing the release block label then.

**Comment by mykaul** [comment]
@bhalevy - should this also move to 6.1?

**Comment by bhalevy** [comment]
> @bhalevy - should this also move to 6.1?

yes, and it may be dropped altogether when we prove we don't it, once https://github.com/scylladb/scylladb/issues/17572 is done, based on https://github.com/scylladb/scylladb/issues/16129

**Comment by mykaul** [comment]
@bhalevy - what's the latest with this item?

**Comment by bhalevy** [comment]
> @bhalevy - what's the latest with this item?

We came to the conclusion that `nodetool rebuild` does not have to have special support for tablets, as replication factor changes trigger automatic rebuild of tablets.
This should be documented, and tests that run nodetool rebuild with tablets need to be adjusted, and just make sure nothing bad happens when running it in those conditions.

That said, we do need to add support for changing replication factor from 0 to N when adding a data center, and from N to 0 before removing a data center, and let the rf change code automate the replica building.

**Comment by bhalevy** [comment]
raising priority to P1 since `nodetool rebuild` silently ignores tablet keyspaces and this is misleading since the casual user assumes it rebuilt all keyspaces on the node.

At the very least we should indicate an error if there are keyspaces with tablets so that the user can run repair instead.

Since the tablets on the node that is being rebuilt are subject to migration, the user should repair all nodes in the datacenter (although current tablets migration code confines migration to the same rack, but it may change in the future without notice, so better recommend to repair all nodes in the datacenter).

**Comment by bhalevy** [comment]
> The issue was changed from "Support rebuild with tablets" to a warning: "recovering from local data loss on this node requires running repair on all nodes in the datacenter".
> 
> Can we close this issue and open a new one to support rebuild with tablets?

What would be different in a new issue from this one?

**Comment by mykaul** [comment]
I'm not sure why it's on 'Backlog' if it's a P1.

**Comment by mykaul** [comment]
@regevran - is this in progress? Who's actively working on it? What do we need here?

**Comment by regevran** [comment]
@mykaul - This is not in progress.
Changing RF from 0 to N and from N to 0 (#17579 and #17572) is be the technique to "rebuild" with tablets.

**Comment by mykaul** [comment]
> [@mykaul](https://github.com/mykaul) - This is not in progress. Changing RF from 0 to N and from N to 0 ([#17579](https://github.com/scylladb/scylladb/issues/17579) and [#17572](https://github.com/scylladb/scylladb/issues/17572)) is be the technique to "rebuild" with tablets.

If it's not, we better fix the status in the various projects this issue is linked to.

**Comment by bhalevy** [comment]
Reassigning to @denesb as his team is working on https://github.com/scylladb/scylladb/issues/17572 and https://github.com/scylladb/scylladb/issues/17579

**Comment by Deexie** [comment]
I'm confused with the comments: do we want `nodetool rebuild` to support tablets or do we assume that this is done by replication factor changes and this issue is an umbrella issue for 0->N and N->0?

@bhalevy @regevran

**Comment by avikivity** [comment]
Let's list the cases for `nodetool rebuild` with vnodes and see if we they are covered by tablets.

1. Changing replication factor - covered by the ALTER TABLE work
2. Fixing up a node after its data was wiped (say due to a faulty disk) - risky, and better done by killing the node and adding a new one instead (perhaps reusing the hardware)
3. any more?

If we don't have a candidate for 3, we can drop support for `nodetool rebuild` for tablets.

**Comment by Deexie** [comment]
> Let's list the cases for `nodetool rebuild` with vnodes and see if we they are covered by tablets.
> 
>     1. Changing replication factor - covered by the ALTER TABLE work
> 
>     2. Fixing up a node after its data was wiped (say due to a faulty disk) - risky, and better done by killing the node and adding a new one instead (perhaps reusing the hardware)
> 
>     3. any more?
> 
> 
> If we don't have a candidate for 3, we can drop support for `nodetool rebuild` for tablets.

So, I asked a few people and cloud and devops use it only after adding a new DC. CX uses `nodetool rebuild` also in a case, when they replace a node with `auto_bootstrap` off.  This isn't relevant for tablets, is it?

**Comment by avikivity** [comment]
> > Let's list the cases for `nodetool rebuild` with vnodes and see if we they are covered by tablets.
> > ```
> > 1. Changing replication factor - covered by the ALTER TABLE work
> > 
> > 2. Fixing up a node after its data was wiped (say due to a faulty disk) - risky, and better done by killing the node and adding a new one instead (perhaps reusing the hardware)
> > 
> > 3. any more?
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If we don't have a candidate for 3, we can drop support for `nodetool rebuild` for tablets.
> 
> So, I asked a few people and cloud and devops use it only after adding a new DC. CX uses `nodetool rebuild` also in a case, when they replace a node with `auto_bootstrap` off. This isn't relevant for tablets, is it?

It is not relevant for tablets.

I'm surprised we haven't removed auto_bootstrap=false.

**Comment by bhalevy** [comment]
> Let's list the cases for `nodetool rebuild` with vnodes and see if we they are covered by tablets.
> 
> 1. Changing replication factor - covered by the ALTER TABLE work
> 2. Fixing up a node after its data was wiped (say due to a faulty disk) - risky, and better done by killing the node and adding a new one instead (perhaps reusing the hardware)
> 3. any more?
> 
> If we don't have a candidate for 3, we can drop support for `nodetool rebuild` for tablets.

We already did, didn't we?

We use `database::get_non_local_strategy_keyspaces_erms()` in storage_service, and it returns only vnode non-local keyspaces.
See https://github.com/scylladb/scylladb/blob/d9aec89c4ee70282ea4e980a5e998c2dd032bd4b/replica/database.cc#L1232

**Comment by Deexie** [comment]
> > Let's list the cases for `nodetool rebuild` with vnodes and see if we they are covered by tablets.
> > 
> > 1. Changing replication factor - covered by the ALTER TABLE work
> > 2. Fixing up a node after its data was wiped (say due to a faulty disk) - risky, and better done by killing the node and adding a new one instead (perhaps reusing the hardware)
> > 3. any more?
> > 
> > If we don't have a candidate for 3, we can drop support for `nodetool rebuild` for tablets.
> 
> We already did, didn't we?
> 
> We use `database::get_non_local_strategy_keyspaces_erms()` in storage_service, and it returns only vnode non-local keyspaces. See
> 
> [scylladb/replica/database.cc](https://github.com/scylladb/scylladb/blob/d9aec89c4ee70282ea4e980a5e998c2dd032bd4b/replica/database.cc#L1232)
> 
> Line 1232 in [d9aec89](/scylladb/scylladb/commit/d9aec89c4ee70282ea4e980a5e998c2dd032bd4b)
>  if (rs.get_type() != locator::replication_strategy_type::local && !rs.is_per_table()) {

I think Avi meant "drop a plan to add support"

**Comment by denesb** [comment]
So the only thing that remains to do here is to ensure that the rebuild API and nodetool reject rebuild attempt for tablet keyspaces/tables.

**Comment by Deexie** [comment]
> So the only thing that remains to do here is to ensure that the rebuild API and nodetool reject rebuild attempt for tablet keyspaces/tables.

I just realized that we do not specify keyspace for rebuild, so there is no point in rejecting anything - we want to rebuild vnode keyspaces even though there are tablet keyspaces, too. I think current warning:
```
slogger.warn("Rebuild is not supported for the following tablets-enabled keyspaces: {}: "
                    "Rebuild is not required for tablets-enabled keyspace after increasing replication factor. "
                    "However, recovering from local data loss on this node requires running repair on all nodes in the datacenter", tablets_keyspaces);
```
is sufficient.

**Comment by pdbossman** [comment]
Today, when we add a DC, we alter keyspaces, then run rebuild, when rebuild completes for all nodes in the DC - user can switch application traffic to the new DC.

How will the user know that data migration for tablets from source to target DC is complete and it is safe to switch application traffic to the new DC?

**Comment by pdbossman** [comment]
The rebuild will complete, but tablets may still be migrating and switching application traffic early may produce incorrect results.

**Comment by Deexie** [comment]
> Today, when we add a DC, we alter keyspaces, then run rebuild, when rebuild completes for all nodes in the DC - user can switch application traffic to the new DC.
> 
> How will the user know that data migration for tablets from source to target DC is complete and it is safe to switch application traffic to the new DC?

The data migration will be done when `ALTER KEYSPACE` returns (unless it times out).

**Comment by pdbossman** [comment]
> > Today, when we add a DC, we alter keyspaces, then run rebuild, when rebuild completes for all nodes in the DC - user can switch application traffic to the new DC.
> > How will the user know that data migration for tablets from source to target DC is complete and it is safe to switch application traffic to the new DC?
> 
> The data migration will be done when `ALTER KEYSPACE` returns (unless it times out).

@ptrsmrn ^^^ this conflicts with your IIRC that ALTER KEYSPACE returns and only schedules data movement.  

I think we need a definitive answer here and clear instructions for add a DC process - since when it's safe to use the new DC when there are tablets vs vnode vs both is different.

**Comment by ptrsmrn** [comment]
> > > Today, when we add a DC, we alter keyspaces, then run rebuild, when rebuild completes for all nodes in the DC - user can switch application traffic to the new DC.
> > > How will the user know that data migration for tablets from source to target DC is complete and it is safe to switch application traffic to the new DC?
> > 
> > 
> > The data migration will be done when `ALTER KEYSPACE` returns (unless it times out).
> 
> [@ptrsmrn](https://github.com/ptrsmrn) ^^^ this conflicts with your IIRC that ALTER KEYSPACE returns and only schedules data movement.
> 
> I think we need a definitive answer here and clear instructions for add a DC process - since when it's safe to use the new DC when there are tablets vs vnode vs both is different.

@pdbossman I don't think the code changed that much since ALTERing tablets KEYSPACEs was implemented, and back then [ALTER only scheduled tablets transitions](https://github.com/scylladb/scylladb/blob/master/service/topology_coordinator.cc#L972) and [didn't wait for their completion](https://github.com/scylladb/scylladb/blame/master/cql3/query_processor.cc#L1078).

**Comment by Deexie** [comment]
> > > > Today, when we add a DC, we alter keyspaces, then run rebuild, when rebuild completes for all nodes in the DC - user can switch application traffic to the new DC.
> > > > How will the user know that data migration for tablets from source to target DC is complete and it is safe to switch application traffic to the new DC?
> > > 
> > > 
> > > The data migration will be done when `ALTER KEYSPACE` returns (unless it times out).
> > 
> > 
> > [@ptrsmrn](https://github.com/ptrsmrn) ^^^ this conflicts with your IIRC that ALTER KEYSPACE returns and only schedules data movement.
> > I think we need a definitive answer here and clear instructions for add a DC process - since when it's safe to use the new DC when there are tablets vs vnode vs both is different.
> 
> [@pdbossman](https://github.com/pdbossman) I don't think the code changed that much since ALTERing tablets KEYSPACEs was implemented, and back then [ALTER only scheduled tablets transitions](https://github.com/scylladb/scylladb/blob/master/service/topology_coordinator.cc#L972) and [didn't wait for their completion](https://github.com/scylladb/scylladb/blame/master/cql3/query_processor.cc#L1078).

Don't we wait for the tablets transitions in:
```
co_await remote_.get().ss.wait_for_topology_not_busy();
```

**Comment by ptrsmrn** [comment]
> > > > > Today, when we add a DC, we alter keyspaces, then run rebuild, when rebuild completes for all nodes in the DC - user can switch application traffic to the new DC.
> > > > > How will the user know that data migration for tablets from source to target DC is complete and it is safe to switch application traffic to the new DC?
> > > > 
> > > > 
> > > > The data migration will be done when `ALTER KEYSPACE` returns (unless it times out).
> > > 
> > > 
> > > [@ptrsmrn](https://github.com/ptrsmrn) ^^^ this conflicts with your IIRC that ALTER KEYSPACE returns and only schedules data movement.
> > > I think we need a definitive answer here and clear instructions for add a DC process - since when it's safe to use the new DC when there are tablets vs vnode vs both is different.
> > 
> > 
> > [@pdbossman](https://github.com/pdbossman) I don't think the code changed that much since ALTERing tablets KEYSPACEs was implemented, and back then [ALTER only scheduled tablets transitions](https://github.com/scylladb/scylladb/blob/master/service/topology_coordinator.cc#L972) and [didn't wait for their completion](https://github.com/scylladb/scylladb/blame/master/cql3/query_processor.cc#L1078).
> 
> Don't we wait for the tablets transitions in:
> 
> ```
> co_await remote_.get().ss.wait_for_topology_not_busy();
> ```

@Deexie To the best of my understanding, topology stops being busy once it completes processing ALTER request, and what ALTER request processing results in are, among others, tablets mutations. Tablets mutations are picked and processed by tablets load balancer independently of ALTER req.

**Comment by bhalevy** [comment]
> > > > > > Today, when we add a DC, we alter keyspaces, then run rebuild, when rebuild completes for all nodes in the DC - user can switch application traffic to the new DC.
> > > > > > How will the user know that data migration for tablets from source to target DC is complete and it is safe to switch application traffic to the new DC?
> > > > > 
> > > > > 
> > > > > The data migration will be done when `ALTER KEYSPACE` returns (unless it times out).
> > > > 
> > > > 
> > > > [@ptrsmrn](https://github.com/ptrsmrn) ^^^ this conflicts with your IIRC that ALTER KEYSPACE returns and only schedules data movement.
> > > > I think we need a definitive answer here and clear instructions for add a DC process - since when it's safe to use the new DC when there are tablets vs vnode vs both is different.
> > > 
> > > 
> > > [@pdbossman](https://github.com/pdbossman) I don't think the code changed that much since ALTERing tablets KEYSPACEs was implemented, and back then [ALTER only scheduled tablets transitions](https://github.com/scylladb/scylladb/blob/master/service/topology_coordinator.cc#L972) and [didn't wait for their completion](https://github.com/scylladb/scylladb/blame/master/cql3/query_processor.cc#L1078).
> > 
> > 
> > Don't we wait for the tablets transitions in:
> > ```
> > co_await remote_.get().ss.wait_for_topology_not_busy();
> > ```
> 
> [@Deexie](https://github.com/Deexie) To the best of my understanding, topology stops being busy once it completes processing ALTER request, and what ALTER request processing results in are, among others, tablets mutations. Tablets mutations are picked and processed by tablets load balancer independently of ALTER req.

@tgrabiec please clarify too, refering to @pdbossman 's question above. When is it safe for the application to use a new DC after ALTER KEYSPACE increasing rhe replication factor in this DC.

I'd also add, where the app uses LOCAL_QUORUM for reads.

**Comment by avikivity** [comment]
When ALTER KEYSPACE returns, the data has been fully replicated.

We need to improve the timeout code and cqlsh so that we don't get timeouts for operations that can take days.

We also need to document a way to get this information if we did time out (or the cqlsh session was lost). Issuing a `DESCRIBE KEYSPACE` and seeing the new datacenter there works, but if it's not there, we can't tell if the operation is still in progress or failed. I guess the task manager will tell us, but with which node?

Maybe it's enough to look at system.topology_requests.

Anyway, the important thing is to document something.

**Comment by avikivity** [comment]
> When ALTER KEYSPACE returns, the data has been fully replicated.

To be clear, that's only true with tablets. With vnodes, you have to go through the rebuild dance.

**Comment by Deexie** [comment]
> We also need to document a way to get this information if we did time out (or the cqlsh session was lost). Issuing a `DESCRIBE KEYSPACE` and seeing the new datacenter there works

This is not true. A change to schema is done together with rebuild schedules, so `DESCRIBE SCHEMA` will tell us if the `ALTER` has started, but not if it has finished.

> , but if it's not there, we can't tell if the operation is still in progress or failed. 

We can only tell that it has started. Additionally, we will never learn that the operation failed. See: https://github.com/scylladb/scylladb/issues/24062.

> I guess the task manager will tell us, but with which node?

Not really. With task manager we can see running rebuilds (visible on each node) and conclude based on that whether the `ALTER` is finished. We don't have a task for `ALTER` itself, but even if we had - global tasks don't store data, they utilize external info (like tables) to present the operation status. But we don't keep that info anywhere.

> 
> Maybe it's enough to look at system.topology_requests.

A request will be done as soon as the rebuilds are scheduled.


(My comment is relevant to current version, after 0->N it will be changed.)

**Comment by pdbossman** [comment]
> When ALTER KEYSPACE returns, the data has been fully replicated.
> 
> We need to improve the timeout code and cqlsh so that we don't get timeouts for operations that can take days.
> 
> We also need to document a way to get this information if we did time out (or the cqlsh session was lost). Issuing a `DESCRIBE KEYSPACE` and seeing the new datacenter there works, but if it's not there, we can't tell if the operation is still in progress or failed. I guess the task manager will tell us, but with which node?
> 
> Maybe it's enough to look at system.topology_requests.
> 
> Anyway, the important thing is to document something.

Additionally, any operation that can take days - it's likely we're going to have to interrupt it for operational reasons.    We'll probably want to run the ALTERs from the target nodes to make sure if we need to restart a node, it's not an application serving one.

Remember to run cqlsh in tmux/screen.

**Comment by tarzanek** [comment]
so what will happen if ALTER will be interrupted when cqlsh running it was shot down for whatever reason? (high TB clusters do really run rebuilds for few days)

**Comment by bhalevy** [comment]
> so what will happen if ALTER will be interrupted when cqlsh running it was shot down for whatever reason? (high TB clusters do really run rebuilds for few days)

Tablet rebuild continues once committed regardless of ALTER KEYSPACE, as documented.

**Comment by tgrabiec** [comment]
> > > > > > > Today, when we add a DC, we alter keyspaces, then run rebuild, when rebuild completes for all nodes in the DC - user can switch application traffic to the new DC.
> > > > > > > How will the user know that data migration for tablets from source to target DC is complete and it is safe to switch application traffic to the new DC?
> > > > > > 
> > > > > > 
> > > > > > The data migration will be done when `ALTER KEYSPACE` returns (unless it times out).
> > > > > 
> > > > > 
> > > > > [@ptrsmrn](https://github.com/ptrsmrn) ^^^ this conflicts with your IIRC that ALTER KEYSPACE returns and only schedules data movement.
> > > > > I think we need a definitive answer here and clear instructions for add a DC process - since when it's safe to use the new DC when there are tablets vs vnode vs both is different.
> > > > 
> > > > 
> > > > [@pdbossman](https://github.com/pdbossman) I don't think the code changed that much since ALTERing tablets KEYSPACEs was implemented, and back then [ALTER only scheduled tablets transitions](https://github.com/scylladb/scylladb/blob/master/service/topology_coordinator.cc#L972) and [didn't wait for their completion](https://github.com/scylladb/scylladb/blame/master/cql3/query_processor.cc#L1078).
> > > 
> > > 
> > > Don't we wait for the tablets transitions in:
> > > ```
> > > co_await remote_.get().ss.wait_for_topology_not_busy();
> > > ```
> > 
> > 
> > [@Deexie](https://github.com/Deexie) To the best of my understanding, topology stops being busy once it completes processing ALTER request, and what ALTER request processing results in are, among others, tablets mutations. Tablets mutations are picked and processed by tablets load balancer independently of ALTER req.
> 
> [@tgrabiec](https://github.com/tgrabiec) please clarify too, refering to [@pdbossman](https://github.com/pdbossman) 's question above. When is it safe for the application to use a new DC after ALTER KEYSPACE increasing rhe replication factor in this DC.
> 
> I'd also add, where the app uses LOCAL_QUORUM for reads.

The idea was that it's safe to use a new DC after ALTER succeeds. There's https://github.com/scylladb/scylladb/issues/24062, which applies to the case of concurrent removenode, in which case success doesn't mean RF was increased for all tablets. When in doubt, or if ALTER failed, you have to check the status manually: `POST /storage_service/quiesce_topology` to wait for active transitions and then check `system.load_per_node` to see if new DC has expected tablet replica count.

**Comment by avikivity** [comment]
> > We also need to document a way to get this information if we did time out (or the cqlsh session was lost). Issuing a `DESCRIBE KEYSPACE` and seeing the new datacenter there works
> 
> This is not true. A change to schema is done together with rebuild schedules, so `DESCRIBE SCHEMA` will tell us if the `ALTER` has started, but not if it has finished.
> 
> > , but if it's not there, we can't tell if the operation is still in progress or failed.
> 
> We can only tell that it has started. Additionally, we will never learn that the operation failed. See: [#24062](https://github.com/scylladb/scylladb/issues/24062).
> 
> > I guess the task manager will tell us, but with which node?
> 
> Not really. With task manager we can see running rebuilds (visible on each node) and conclude based on that whether the `ALTER` is finished. We don't have a task for `ALTER` itself, but even if we had - global tasks don't store data, they utilize external info (like tables) to present the operation status. But we don't keep that info anywhere.
> 
> > Maybe it's enough to look at system.topology_requests.
> 
> A request will be done as soon as the rebuilds are scheduled.
> 
> (My comment is relevant to current version, after 0->N it will be changed.)

Well, let's discuss how it looks after the 0->N change, that's what's important.

**Comment by avikivity** [comment]
> > When ALTER KEYSPACE returns, the data has been fully replicated.
> > We need to improve the timeout code and cqlsh so that we don't get timeouts for operations that can take days.
> > We also need to document a way to get this information if we did time out (or the cqlsh session was lost). Issuing a `DESCRIBE KEYSPACE` and seeing the new datacenter there works, but if it's not there, we can't tell if the operation is still in progress or failed. I guess the task manager will tell us, but with which node?
> > Maybe it's enough to look at system.topology_requests.
> > Anyway, the important thing is to document something.
> 
> Additionally, any operation that can take days - it's likely we're going to have to interrupt it for operational reasons.

We'll need to add cancellation on top.

**Comment by Deexie** [comment]
> Well, let's discuss how it looks after the 0->N change, that's what's important.

Ok, after the 0->N change:

>  We also need to document a way to get this information if we did time out (or the cqlsh session was lost). Issuing a `DESCRIBE KEYSPACE` and seeing the new datacenter there works

It is true. We will see the previous replication if rf change is in progress (even if replica count is temporarily different) and target replication when rf change is done.

> , but if it's not there, we can't tell if the operation is still in progress or failed.

We will be able to tell so based on the content of a new column in system.topology (has content if rf change in progress) or system.topology_requests.

 > I guess the task manager will tell us, but with which node?

No, task manager won't tell us. Contrary to current state, we won't be able to conclude about the progress of rf change based on ongoing rebuilds. (Rebuilds won't be scheduled all at once; they might mix with migrations etc.)

We may add a task to cover rf change, but that won't be a part of 0->N PR itself.

  > Maybe it's enough to look at system.topology_requests.

Yes.



To sum up: after 0->N change:
- describe shows new replication if rf change is done and previous otherwise;
- system.topology_requests shows whether rf change succeed, failed, or is in progress;
- we are able to conclude about the progress of rf change based on system.topology.

**Comment by Deexie** [comment]
> > > When ALTER KEYSPACE returns, the data has been fully replicated.
> > > We need to improve the timeout code and cqlsh so that we don't get timeouts for operations that can take days.
> > > We also need to document a way to get this information if we did time out (or the cqlsh session was lost). Issuing a `DESCRIBE KEYSPACE` and seeing the new datacenter there works, but if it's not there, we can't tell if the operation is still in progress or failed. I guess the task manager will tell us, but with which node?
> > > Maybe it's enough to look at system.topology_requests.
> > > Anyway, the important thing is to document something.
> > 
> > 
> > Additionally, any operation that can take days - it's likely we're going to have to interrupt it for operational reasons.
> 
> We'll need to add cancellation on top.

0->N rf change is designed in a way that supports cancellation. I think task manager tasks would be the clearest for that.

**Comment by pdbossman** [comment]
Some additional operational context.
For rebuild add a dc, with new DCs sometimes distant from source, we would rebuild rack 1, then source rack 2 and 3 from target DC, then follow with full repair.  The low-latency within target rack was considerably faster, and since we hadn't flipped apps over, rebuilding 2nd and 3rd racks was no impact to source cluster, we could run more concurrently, etc.

Not sure what sort of optimizations we consider in this regard.

**Comment by avikivity** [comment]
> > Well, let's discuss how it looks after the 0->N change, that's what's important.
> 
> Ok, after the 0->N change:
> 
> > We also need to document a way to get this information if we did time out (or the cqlsh session was lost). Issuing a `DESCRIBE KEYSPACE` and seeing the new datacenter there works
> 
> It is true. We will see the previous replication if rf change is in progress (even if replica count is temporarily different) and target replication when rf change is done.


Just for fun, we can add the pending replication layout in a CQL comment (`-- blah blah`).

> 
> > , but if it's not there, we can't tell if the operation is still in progress or failed.
> 
> We will be able to tell so based on the content of a new column in system.topology (has content if rf change in progress) or system.topology_requests.
> 
> > I guess the task manager will tell us, but with which node?
> 
> No, task manager won't tell us. Contrary to current state, we won't be able to conclude about the progress of rf change based on ongoing rebuilds. (Rebuilds won't be scheduled all at once; they might mix with migrations etc.)
> 
> We may add a task to cover rf change, but that won't be a part of 0->N PR itself.
> 

Ok.

> > Maybe it's enough to look at system.topology_requests.
> 
> Yes.
> 
> To sum up: after 0->N change:
> 
> * describe shows new replication if rf change is done and previous otherwise;
> * system.topology_requests shows whether rf change succeed, failed, or is in progress;
> * we are able to conclude about the progress of rf change based on system.topology.

**Comment by denesb** [comment]
This is a duplicate of https://scylladb.atlassian.net/browse/SCYLLADB-103. According to @Deexie there will be a single [PR](https://github.com/scylladb/scylladb/pull/24421) adding support for both 0->N (rebuild DC) and N->0 (decomission DC). This includes tablet rebuild. Any documentation update to explain how rebuild works for tablets will also be handled by said PR.

Closing this as wontfix, as github doesn't allow close as dup of JIRA issue.

**Comment by piodul** [comment]
Reopening, some dtests got reenabled. They fail because they were not adjusted to rf-rack-validity, but considering that the issue is not really solved but moved to Jira, I suppose they should not be allowed to run?

@denesb I think we will be able to close the issue only after we adjust the `issue_open` labels to point to the Jira issue...

**Comment by piodul** [comment]
Example here: https://jenkins.scylladb.com/job/scylla-2025.4/job/next/161/

**Comment by pehala** [comment]
> Reopening, some dtests got reenabled. They fail because they were not adjusted to rf-rack-validity, but considering that the issue is not really solved but moved to Jira, I suppose they should not be allowed to run?
> 
> [@denesb](https://github.com/denesb) I think we will be able to close the issue only after we adjust the `issue_open` labels to point to the Jira issue...

Sadly, yes when moving to Jira, dtest update is needed :/

EDIT: or marking it as complete, I think it wont work with "not planned" but that could be adjusted

**Comment by bhalevy** [comment]
@denesb @pehala can we close this issue (as Not planned, rather than Completed)?

**Comment by pehala** [comment]
> This is a duplicate of https://scylladb.atlassian.net/browse/SCYLLADB-103. According to [@Deexie](https://github.com/Deexie) there will be a single [PR](https://github.com/scylladb/scylladb/pull/24421) adding support for both 0->N (rebuild DC) and N->0 (decomission DC).


> [@denesb](https://github.com/denesb) [@pehala](https://github.com/pehala) can we close this issue (as Not planned, rather than Completed)?

@bhalevy so it is not planned in the end and wont be addressed by the 0->N feature, or the issue here is not planned due to JIRA migration?

**Comment by bhalevy** [comment]
> > This is a duplicate of [scylladb.atlassian.net/browse/SCYLLADB-103](https://scylladb.atlassian.net/browse/SCYLLADB-103). According to [@Deexie](https://github.com/Deexie) there will be a single [PR](https://github.com/scylladb/scylladb/pull/24421) adding support for both 0->N (rebuild DC) and N->0 (decomission DC).
> 
> > [@denesb](https://github.com/denesb) [@pehala](https://github.com/pehala) can we close this issue (as Not planned, rather than Completed)?
> 
> [@bhalevy](https://github.com/bhalevy) so it is not planned in the end and wont be addressed by the 0->N feature, or the issue here is not planned due to JIRA migration?

I'm afraid there's a confusion in terms between nodetool rebuild for tablets and tablet-rebuild which is an internal transition that is used as part of RF increase.

This issue was originally about supporting nodetool rebuild for tablets, but that's not going to happen, and the way I see it, once we get rid of vnodes there would be no need for nodetool rebuild and we can stop stop supporting it all.

The reason is hat nodetool rebuild is used for 2 distinct use-cases:
1) When a node loses its local storage.  In this case, the safest way is to replace the node by itself (or remove it using nodetool removenode and then adding the instance as a bootstrapping node), this works for both vnodes and tablets and the advantage in doing that is that the cluster is aware that it lost data once the node is excluded (after nodetool removenode or when replaced).  It can also be repaired in place using nodetool repair (or nodetool rebuild for vnodes), but until this finished the system is exposed since it might break quorum guarantees - for a example if a write reached replicas A,B and not C, then node B lost its storage, reading in quorum from B and C will see old or no data, while if B is excluded, quorum read will be forced to A and C and will repair the data correctly.

2) The other use case for nodetool rebuild for vnodes is after increasing RF onto a new datacenter/rack.  This is required for vnodes as ALTER KEYSPACE just changes the schema without rebuilding the data and that is required before starting to access the data on the new DC.  It is not needed for tablets as we rebuild the tablet replicas synchronously upon ALTER KEYSPACE with the additional replica.  @denesb's comment about 0->N is to allow 0->N RF change for a new DC that will rebuild all replicas internally as part of the respective ALTER KEYSPACE, but there is no need to make nodetool rebuild work for tablets - only ALTER KEYSPACE.

**Comment by mykaul** [comment]
While 'there's no need' for the 2nd use case:
1. No one knows that there's no need.
2. They expect it to work, even if it does nothing, or you might be breaking some automation ornold habits.

**Comment by bhalevy** [comment]
> While 'there's no need' for the 2nd use case:
> 
> 1. No one knows that there's no need.
> 2. They expect it to work, even if it does nothing, or you might be breaking some automation ornold habits.

nodetool rebuild works for adding a dc in the sense it simply ignores tablets keyspaces and rebuild the replicas of vnodes keyspaces (following an increase of the rf in the added dc).  The end results is that both vnodes and tablets keyspaces replicas are rebuilt in the new dc, so nothing is broken.

After getting rid of vnodes there would no need for nodetool rebuild since it will do nothing.

**Comment by bhalevy** [comment]
I opened https://scylladb.atlassian.net/browse/DTEST-11 to get rid of the dependency on this issue so we can close it as not planned.

**Comment by Deexie** [comment]
Dtest fix is already in, closing
